### ISSUE TITLE
In SimMaxLoss， why rank=rank-1 ？

### ISSUE THREAD

**Issue by YingLv1106** [CVI-SZU/CCAM#7]
class SimMaxLoss(nn.Module):
    def __init__(self, metric='cos', alpha=0.25, reduction='mean'):
        super(SimMaxLoss, self).__init__()
        self.metric = metric
        self.alpha = alpha
        self.reduction = reduction

    def forward(self, embedded_bg):
        """
        :param embedded_fg: [N, C]
        :param embedded_bg: [N, C]
        :return:
        """
        if self.metric == 'l2':
            raise NotImplementedError

        elif self.metric == 'cos':
            sim = cos_simi(embedded_bg, embedded_bg)  #(N,N)
            loss = -torch.log(sim)
            loss[loss < 0] = 0
            _, indices = sim.sort(descending=True, dim=1)
            _, rank = indices.sort(dim=1)
            rank = rank - 1
            rank_weights = torch.exp(-rank.float() * self.alpha)
            loss = loss * rank_weights
        else:
            raise NotImplementedError

        if self.reduction == 'mean':
            return torch.mean(loss)
        elif self.reduction == 'sum':
            return torch.sum(loss)

**Comment by Sierkinhane** [comment]
Hi. please note that the diagonal elements (=1.0) in 'sim' are the product of two same embeddings and they rank 0. rank-1 aims to give weights to other items from 1 (exp(0)=1) to 0. And because log(1)=0, the diagonal elements in 'loss' are 0 and it will not be affected by the rank-1 (0-1=-1).

**Comment by YingLv1106** [comment]
Thanks, I get it. Excluding its own similarity terms, consider the weight of other terms as 0 to 1.

**Comment by Sierkinhane** [comment]
Exactly!
